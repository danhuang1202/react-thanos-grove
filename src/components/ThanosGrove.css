.container {
  position: relative;
  width: 80px;
  height: 80px;
}

.img {
  object-fit: contain;
}

.screen {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  overflow: hidden;
}

.film {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
}
