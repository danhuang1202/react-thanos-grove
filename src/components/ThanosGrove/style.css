.container {
  position: relative;
  width: 80px;
  height: 80px;
  cursor: pointer;
}

.disable {
  cursor: not-allowed;
  pointer-events: none;
}

.img {
  object-fit: contain;
}

.screen {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  overflow: hidden;
}

.film {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  will-change: transform;
}
