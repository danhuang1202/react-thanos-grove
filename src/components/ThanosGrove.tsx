import React, { FunctionComponent, useState, createRef } from 'react'
import cns from 'classnames'
// @ts-ignore
import style from './ThanosGrove.css'

type Props = {
  /** Custom class name */
  className?: string
  /** callback function will run with decimate */
  onDecimation?: () => void
  /** callback function will run with reverse */
  onReverse?: () => void
}

const FOOTAGE = 3480
const FILE_FRAME_TIME = 60
const FILE_FRAME_WIDTH = 80
const FILE_FRAME_COUNT = FOOTAGE / FILE_FRAME_WIDTH
const IMAGE_THANOS_IDLE =
  'https://www.google.com/logos/fnbx/thanos/thanos_idle.png'
const IMAGE_THANOS_SNAP =
  'https://www.google.com/logos/fnbx/thanos/thanos_snap.png'
const IMAGE_THANOS_TIME =
  'https://www.google.com/logos/fnbx/thanos/thanos_time.png'
const AUDIO_THANOS_REVERSE =
  'https://www.google.com/logos/fnbx/thanos/thanos_reverse_sound.mp3'
const AUDIO_THANOS_SNAP =
  'https://www.google.com/logos/fnbx/thanos/thanos_snap_sound.mp3'

const ThanosGrove: FunctionComponent<Props> = ({
  className,
  onDecimation,
  onReverse
}) => {
  const [decimation, setDecimation] = useState(false)
  const [isSnaping, setIsSnaping] = useState(false)
  const filmEl = createRef<HTMLDivElement>()
  const audioEl = createRef<HTMLAudioElement>()

  const clickHandler = () => {
    const target = filmEl.current
    let startTime = null
    console.log('clickHandler', filmEl)
    const updateFrame = timestamp => {
      console.log('updateFrame', filmEl)
      if (!startTime) startTime = timestamp

      const frameIndex = Math.min(
        Math.floor((timestamp - startTime) / FILE_FRAME_TIME),
        FILE_FRAME_COUNT
      )
      const translateX = `-${frameIndex * FILE_FRAME_WIDTH}px`
      target.style.transform = `translateX(${translateX})`

      if (frameIndex < FILE_FRAME_COUNT) {
        requestAnimationFrame(updateFrame)
      } else {
        if (decimation) {
          onReverse && onReverse()
        } else onDecimation && onDecimation()

        target.style.transform = 'translateX(0)'
        setIsSnaping(false)
        setDecimation(!decimation)
      }
    }

    if (audioEl.current) {
      audioEl.current.load()
      audioEl.current.play()
    }
    setIsSnaping(true)
    requestAnimationFrame(updateFrame)
  }

  const audioSource = decimation ? AUDIO_THANOS_REVERSE : AUDIO_THANOS_SNAP
  const filmStyle = {
    width: `${FOOTAGE}px`,
    backgroundImage: `url('${
      decimation ? IMAGE_THANOS_TIME : IMAGE_THANOS_SNAP
    }')`
  }

  return (
    <div className={cns(style.container, className)}>
      <img
        alt="thanos grove"
        width="80px"
        height="80px"
        src={IMAGE_THANOS_IDLE}
        className={style.img}
        hidden={isSnaping}
      />
      <div className={style.screen} onClick={clickHandler}>
        <div ref={filmEl} className={style.film} style={filmStyle} />
        <audio controls ref={audioEl} className={style.audio} hidden>
          <source src={audioSource} type="audio/mpeg" />
        </audio>
      </div>
    </div>
  )
}

export default ThanosGrove
