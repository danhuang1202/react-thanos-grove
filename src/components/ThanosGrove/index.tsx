import React, { FunctionComponent, useState, useRef } from 'react'
import cns from 'classnames'
import useRequestAnimationFrame from '../hooks/useRequestAnimationFrame'
// @ts-ignore
import style from './style.css'

type Props = {
  /** Custom class name */
  className?: string
  /** callback function will run with decimate */
  onDecimation?: () => void
  /** callback function will run with reverse */
  onReverse?: () => void
}

const FOOTAGE = 3840
const FILE_FRAME_TIME = 32
const FILE_FRAME_WIDTH = 80
const FILE_FRAME_COUNT = FOOTAGE / FILE_FRAME_WIDTH
const IMAGE_THANOS_IDLE =
  'https://www.google.com/logos/fnbx/thanos/thanos_idle.png'
const IMAGE_THANOS_SNAP =
  'https://www.google.com/logos/fnbx/thanos/thanos_snap.png'
const IMAGE_THANOS_TIME =
  'https://www.google.com/logos/fnbx/thanos/thanos_time.png'
const AUDIO_THANOS_REVERSE =
  'https://www.google.com/logos/fnbx/thanos/thanos_reverse_sound.mp3'
const AUDIO_THANOS_SNAP =
  'https://www.google.com/logos/fnbx/thanos/thanos_snap_sound.mp3'

const ThanosGrove: FunctionComponent<Props> = ({
  className,
  onDecimation,
  onReverse
}) => {
  const [decimation, setDecimation] = useState(false)
  const [isSnaping, setIsSnaping] = useState(false)
  const filmEl = useRef<HTMLDivElement>()
  const audioEl = useRef<HTMLAudioElement>()
  const duration = FILE_FRAME_TIME * FILE_FRAME_COUNT
  const callback = elapsedTime => {
    const frameIndex = Math.min(
      Math.floor(elapsedTime / FILE_FRAME_TIME),
      FILE_FRAME_COUNT
    )
    const translateX = `-${frameIndex * FILE_FRAME_WIDTH}px`
    filmEl.current.style.transform = `translateX(${translateX})`

    if (elapsedTime >= duration) {
      if (decimation) {
        onReverse && onReverse()
      } else {
        onDecimation && onDecimation()
      }

      setIsSnaping(false)
      setDecimation(!decimation)
    }
  }

  const { isActive, start } = useRequestAnimationFrame({
    disable: true,
    duration,
    callback
  })

  const clickHandler = () => {
    if (isActive) {
      return
    }

    if (audioEl.current) {
      audioEl.current.load()
      audioEl.current.play()
    }
    filmEl.current.style.transform = 'translateX(0)'
    setIsSnaping(true)
    start()
  }

  const audioSource = decimation ? AUDIO_THANOS_REVERSE : AUDIO_THANOS_SNAP
  const filmStyle = {
    width: `${FOOTAGE}px`,
    backgroundImage: `url('${
      decimation ? IMAGE_THANOS_TIME : IMAGE_THANOS_SNAP
    }')`
  }

  return (
    <div
      data-testid="grove"
      className={cns(style.container, className)}
      onClick={clickHandler}
    >
      <img
        alt="thanos grove"
        width="80px"
        height="80px"
        src={IMAGE_THANOS_IDLE}
        className={style.img}
        hidden={isSnaping}
      />
      <div className={style.screen}>
        <div ref={filmEl} className={style.film} style={filmStyle} />
        <audio controls ref={audioEl} className={style.audio} hidden>
          <source src={audioSource} type="audio/mpeg" />
        </audio>
      </div>
    </div>
  )
}

export default ThanosGrove
