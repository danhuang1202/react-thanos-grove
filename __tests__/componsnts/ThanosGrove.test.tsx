import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import ThanosGrove from '@components/ThanosGrove'

let time = Date.now()
jest.spyOn(window, 'requestAnimationFrame').mockImplementation(callback => {
  time += 60000
  callback(time)
  return time
})

describe('ThanosGrove', () => {
  it('[sanpshot] before decimation', () => {
    const { container } = render(<ThanosGrove />)

    expect(container).toMatchSnapshot()
  })

  it('[sanpshot] after decimation', () => {
    jest.useFakeTimers()

    const { container, getByTestId } = render(<ThanosGrove />)

    fireEvent.click(getByTestId('grove'))

    jest.advanceTimersByTime(60000)

    expect(container).toMatchSnapshot()

    jest.clearAllTimers()
  })

  it('should call handler function', () => {
    const onDecimation = jest.fn(() => {})
    const onReverse = jest.fn(() => {})

    const { getByTestId } = render(
      <ThanosGrove onDecimation={onDecimation} onReverse={onReverse} />
    )

    fireEvent.click(getByTestId('grove'))
    jest.advanceTimersByTime(60000)

    expect(onDecimation).toHaveBeenCalled()

    fireEvent.click(getByTestId('grove'))
    jest.advanceTimersByTime(60000)

    expect(onReverse).toHaveBeenCalled()
  })
})
