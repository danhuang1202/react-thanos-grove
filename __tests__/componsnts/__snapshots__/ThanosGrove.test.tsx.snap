// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThanosGrove [sanpshot] after decimation 1`] = `
<div>
  <div
    class="ThanosGrove_container_1ummv"
    data-testid="grove"
  >
    <img
      alt="thanos grove"
      class="ThanosGrove_img_2AS2q"
      height="80px"
      src="https://www.google.com/logos/fnbx/thanos/thanos_idle.png"
      width="80px"
    />
    <div
      class="ThanosGrove_screen_32q9N"
    >
      <div
        class="ThanosGrove_film_2lKBj"
        style="width: 3840px; background-image: url(https://www.google.com/logos/fnbx/thanos/thanos_time.png); transform: translateX(-3840px);"
      />
      <audio
        controls=""
        hidden=""
      >
        <source
          src="https://www.google.com/logos/fnbx/thanos/thanos_reverse_sound.mp3"
          type="audio/mpeg"
        />
      </audio>
    </div>
  </div>
</div>
`;

exports[`ThanosGrove [sanpshot] before decimation 1`] = `
<div>
  <div
    class="ThanosGrove_container_1ummv"
    data-testid="grove"
  >
    <img
      alt="thanos grove"
      class="ThanosGrove_img_2AS2q"
      height="80px"
      src="https://www.google.com/logos/fnbx/thanos/thanos_idle.png"
      width="80px"
    />
    <div
      class="ThanosGrove_screen_32q9N"
    >
      <div
        class="ThanosGrove_film_2lKBj"
        style="width: 3840px; background-image: url(https://www.google.com/logos/fnbx/thanos/thanos_snap.png);"
      />
      <audio
        controls=""
        hidden=""
      >
        <source
          src="https://www.google.com/logos/fnbx/thanos/thanos_snap_sound.mp3"
          type="audio/mpeg"
        />
      </audio>
    </div>
  </div>
</div>
`;
